draco/tests/test_hard.py

```diff
@@ -2008,6 +2008,631 @@ def test_col_no_x():
+        list_violations(
+            p
+            + """
+    entity(encoding,2,4).
+    attribute((encoding,channel),4,y).
+    attribute((encoding,aggregate),4,max).
+    attribute((encoding,stack),4,normalize).
+    """
+        )
+        == ["stack_without_summative_agg"]
+    )
+
+
+def test_no_stack_with_bar_area_discrete_color():
+    b = hard.blocks["no_stack_with_bar_area_discrete_color"]
+    assert isinstance(b, Block)
+    p = b.program
+
+    assert no_violations(
+        p
+        + """
+    entity(mark,0,1).
+    attribute((mark,type),1,bar).
+    entity(encoding,1,3).
+    attribute((encoding,channel),3,y).
+    attribute((encoding,aggregate),3,count).
+    attribute((encoding,stack),3,zero).
+    entity(encoding,1,4).
+    attribute((encoding,channel),4,color).
+    attribute((encoding,field),4,condition).
+    entity(scale,0,6).
+    attribute((scale,channel),6,y).
+    attribute((scale,type),6,linear).
+    attribute((scale,zero),6,true).
+    entity(scale,0,7).
+    attribute((scale,channel),7,color).
+    attribute((scale,type),7,categorical).
+    """
+    )
+
+    # bar/area mark with discrete color, must stack
+    assert (
+        list_violations(
+            p
+            + """
+    entity(mark,0,1).
+    attribute((mark,type),1,bar).
+    entity(encoding,1,3).
+    attribute((encoding,channel),3,y).
+    attribute((encoding,aggregate),3,count).
+    entity(encoding,1,4).
+    attribute((encoding,channel),4,color).
+    attribute((encoding,field),4,condition).
+    entity(scale,0,6).
+    attribute((scale,channel),6,y).
+    attribute((scale,type),6,linear).
+    attribute((scale,zero),6,true).
+    entity(scale,0,7).
+    attribute((scale,channel),7,color).
+    attribute((scale,type),7,categorical).
+    """
+        )
+        == ["no_stack_with_bar_area_discrete_color"]
+    )
+
+
+def test_stack_without_discrete_color_or_detail():
+    b = hard.blocks["stack_without_discrete_color_or_detail"]
+    assert isinstance(b, Block)
+    p = b.program
+
+    assert no_violations(
+        p
+        + """
+    entity(mark,0,1).
+    attribute((mark,type),1,bar).
+    entity(encoding,1,2).
+    attribute((encoding,channel),2,x).
+    attribute((encoding,field),2,temperature).
+    attribute((encoding,binning),2,10).
+    entity(encoding,1,3).
+    attribute((encoding,channel),3,y).
+    attribute((encoding,aggregate),3,count).
+    attribute((encoding,stack),3,zero).
+    entity(encoding,1,4).
+    attribute((encoding,channel),4,color).
+    attribute((encoding,field),4,condition).
+    entity(scale,0,5).
+    attribute((scale,channel),5,x).
+    attribute((scale,type),5,linear).
+    entity(scale,0,6).
+    attribute((scale,channel),6,y).
+    attribute((scale,type),6,linear).
+    attribute((scale,zero),6,true).
+    entity(scale,0,7).
+    attribute((scale,channel),7,color).
+    attribute((scale,type),7,categorical).
+    """
+    )
+
+    assert no_violations(
+        p
+        + """
+    entity(mark,0,1).
```

Comment:
maybe we can try dict format with such long examples

Reply:
I trimmed the examples to make them shorter